test/plugin/test_base.rb

```diff
@@ -146,4 +146,134 @@ class FluentPluginBaseTest::DummyPlugin2 < Fluent::Plugin::TestBase
       end
     end
   end
+
+  test '`ArgumentError` when `conf` is not `Fluent::Config::Element`' do
+    assert_raise ArgumentError.new('BUG: type of conf must be Fluent::Config::Element, but Hash is passed.') do
+      @p.configure({})
+    end
+  end
+
+  sub_test_case 'system_config.workers value after configure' do
+    sub_test_case 'with <system> workers 3 </system>' do
+      setup do
+        system_config = Fluent::SystemConfig.new
+        system_config.workers = 3
+        stub(Fluent::Engine).system_config { system_config }
+      end
+
+      sub_test_case 'supervisor_mode is true' do
+        setup do
+          stub(Fluent::Engine).supervisor_mode { true }
+          stub(Fluent::Engine).worker_id { -1 }
+        end
+
+        test 'without <worker> directive' do
+          conf = config_element()
+          conf.set_target_worker_ids([])
+          @p.configure(conf)
+          assert{ @p.system_config.workers == 3 }
+        end
```

Comment:
Could you summarize these new tests by using `data`?
It looks like it could be put together into one test.

Example:

https://github.com/fluent/fluentd/blob/fb641737d6bacace19be11cb676fec3fc20f0cf9/test/plugin/test_output.rb#L1077-L1105

Reply:
Changed to test using `data`.

Reply:
Thanks! It is now easier to see!
Certainly, it would be better not to force them all into one test. This looks good.

I'm sorry I said we should put the tests together, but it looks like we need to split the tests up a bit more.
It is generally not desirable to assert different cases in a single test run.

Sorry for writing raw diffs, but how about something like this?
We can use `keep: true` to use the same data in the next test.

```diff
diff --git a/test/plugin/test_base.rb b/test/plugin/test_base.rb
index 7816f1a6..3d4404ba 100644
--- a/test/plugin/test_base.rb
+++ b/test/plugin/test_base.rb
@@ -161,16 +161,9 @@ class BaseTest < Test::Unit::TestCase
       assert{ @p.system_config.workers == data[:expected] }
     end
 
-    def when_supervisor_mode_is_true(data)
+    def stub_supervisor_mode
       stub(Fluent::Engine).supervisor_mode { true }
       stub(Fluent::Engine).worker_id { -1 }
-      assert_system_config_workers_value(data)
-    end
-
-    def when_supervisor_mode_is_false(data)
-      stub(Fluent::Engine).supervisor_mode { false }
-      stub(Fluent::Engine).worker_id { 0 }
-      assert_system_config_workers_value(data)
     end
 
     sub_test_case 'with <system> workers 3 </system>' do
@@ -185,32 +178,40 @@ class BaseTest < Test::Unit::TestCase
         {
           target_worker_ids: [],
           expected: 3
-        }
+        },
+        keep: true
       )
       data(
         'with <worker 0>',
         {
           target_worker_ids: [0],
           expected: 1
-        }
+        },
+        keep: true
       )
       data(
         'with <worker 0-1>',
         {
           target_worker_ids: [0, 1],
           expected: 2
-        }
+        },
+        keep: true
       )
       data(
         'with <worker 0-2>',
         {
           target_worker_ids: [0, 1, 2],
           expected: 3
-        }
+        },
+        keep: true
       )
       test 'system_config.workers value after configure' do
-        when_supervisor_mode_is_true(data)
-        when_supervisor_mode_is_false(data)
+        assert_system_config_workers_value(data)
+      end
+
+      test 'system_config.workers value after configure with supervisor_mode' do
+        stub_supervisor_mode
+        assert_system_config_workers_value(data)
       end
     end
 
@@ -220,18 +221,24 @@ class BaseTest < Test::Unit::TestCase
         {
           target_worker_ids: [],
           expected: 1
-        }
+        },
+        keep: true
       )
       data(
         'with <worker 0>',
         {
           target_worker_ids: [0],
           expected: 1
-        }
+        },
+        keep: true
       )
       test 'system_config.workers value after configure' do
-        when_supervisor_mode_is_true(data)
-        when_supervisor_mode_is_false(data)
+        assert_system_config_workers_value(data)
+      end
+
+      test 'system_config.workers value after configure with supervisor_mode' do
+        stub_supervisor_mode
+        assert_system_config_workers_value(data)
       end
     end
   end
```

Reply:
Thanks for the info.
I was looking for `keep: true`!

Reply:
Thanks for fixing it!

Reply:
I didn't know `keep: true`:
https://www.clear-code.com/blog/2018/12/26.html

Reply:
I know it in that blog too :smile: